File: leetcode-go/Q051-Q100/Q085_Maximal_Rectangle/Q085_Maximal_Rectangle.go
package Q085_Maximal_Rectangle

func maximalRectangle(matrix [][]byte) int {
	if len(matrix) == 0 {
		return 0
	}
	heights := make([]int, len(matrix[0]))
	for i := 0; i < len(matrix[0]); i++ {
		if matrix[0][i] == '0' {
			heights[i] = 0
		} else {
			heights[i] = 1
		}
	}
	heights = append(heights, 0) // add a sentinel to the end

	ans := largestRectangleArea(heights)
	for i := 1; i < len(matrix); i++ {
		for j := 0; j < len(heights)-1; j++ {
			if matrix[i][j] == '0' {
				heights[j] = 0
			} else {
				heights[j] += 1
			}
		}
		ans = max(ans, largestRectangleArea(heights))
	}
	return ans
}

func largestRectangleArea(heights []int) int {
	stack := []int{}
	ans := 0
	for i := 0; i < len(heights); i++ {
		for len(stack) > 0 && heights[stack[len(stack)-1]] >= heights[i] {
			h := heights[stack[len(stack)-1]]
			stack = stack[:len(stack)-1]

			x := -1
			if len(stack) > 0 {
				x = stack[len(stack)-1]
			}
			ans = max(ans, h*(i-x-1))
		}
		stack = append(stack, i)
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: leetcode-java/src/archive/Q001_Q300/Q085_Maximal_Rectangle.java
package archive.Q001_Q300;

import java.util.Stack;

public class Q085_Maximal_Rectangle {

    public int maximalRectangle(char[][] matrix) {

        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }

        int n = matrix.length, m = matrix[0].length;
        int[][] heights = new int[n][m];

        for (int j = 0; j < m; j++) {
            heights[0][j] = matrix[0][j] == '0' ? 0 : 1;
        }

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < m; j++) {
                heights[i][j] = matrix[i][j] == '0' ? 0 : heights[i-1][j]+1;
            }
        }

        int result = 0;
        for(int[] height : heights) {
            int area = 0;
            Stack<Integer> stack = new Stack<>();
            for (int i = 0; i < height.length; i++) {
                while (!stack.isEmpty() && height[i] < height[stack.peek()]) {
                    int pos = stack.pop();
                    int w = stack.isEmpty() ? i : i-stack.peek()-1;
                    int h = height[pos];
                    area = Math.max(area, w*h);
                }
                stack.push(i);
            }
            while (!stack.isEmpty()) {
                int pos = stack.pop();
                int w = stack.isEmpty() ? height.length : height.length-stack.peek()-1;
                int h = height[pos];
                area = Math.max(area, w*h);
            }
            result = Math.max(result, area);
        }

        return result;

    }

}
